src/monmove.cpp

```diff
@@ -1237,6 +1289,87 @@ bool monster::push_to( const tripoint &p, const int boost, const size_t depth )
     return true;
 }
 
+bool monster::take_item_at( const tripoint &p, std::function<std::list<item>::iterator(
+                                map_stack & )> selector )
+{
+    auto items = g->m.i_at( p );
+    auto to_pick_up = selector( items );
+
+    // TODO(sm): no volume/weight limits yet; AFAIK monsters/creatures have no notion of either
+    if( to_pick_up != items.end() ) {
+        // If there is a stack of items on the ground, only pick up as many of them as would
+        // count as "one" item for spawning purposes.
+        int default_charges = to_pick_up->type->charges_default();
+        if( default_charges > 1 && to_pick_up->charges > default_charges ) {
+            to_pick_up->mod_charges( -default_charges );
+            inv.push_back( *to_pick_up );
+            inv.back().charges = default_charges;
+
+            add_msg( m_warning, _( "The %1$s grabs %2$s!" ), name().c_str(),
+                     inv.back().display_name( default_charges ) );
+        } else {
+            std::move( to_pick_up, std::next( to_pick_up ), std::back_inserter( inv ) );
+            items.erase( to_pick_up );
+
+            add_msg( m_warning, _( "The %1$s grabs a %2$s!" ), name().c_str(),
+                     to_pick_up->tname( 1 ) );
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+bool monster::take_food_at( const tripoint &p )
+{
+    bool success = take_item_at( p, [&]( map_stack & items ) {
+        auto items_on_tile = g->m.i_at( p );
+        auto best_food = items.end();
+        float best_food_rating = -1.0;
+
+        for( auto it = items_on_tile.begin(); it != items_on_tile.end(); ++it ) {
+            float rating = rate_food( *it );
+            if( rating > best_food_rating ) {
+                best_food = it;
+            }
+        }
+        return best_food;
+    } );
+
+    if( success ) {
+        // Flee after we've got an item
+        remove_effect( effect_seeking_item );
+        add_effect( effect_run, 50 );
+        moves -= 100;
+    }
+    return success;
+}
+
+bool monster::eat_food_from_inventory()
+{
+    if( inv.empty() ) {
+        return false;
+    }
+
+    // Monsters don't have nutrition, so eating doesn't actually do anything for them yet
+    for( auto it = inv.begin(); it != inv.end(); ++it ) {
+        if( it->is_food() ) {
+            add_msg( _( "The %1$s eats a %2$s." ), name().c_str(), it->tname().c_str() );
```

Comment:
This should probably have a `g->u.sees( *this )` check before. Currently the player character will mysteriously  "sense" when any monster eats.